### ISSUE TITLE
[DX] Rapid resolution change (HW scaler)

### ISSUE THREAD

**Issue by nkast** [MonoGame/MonoGame#3729]
One suggestion to keep your frame rate steady is to adapt resolution on the fly.
This technique doesn't work with monogame. The first frame after changing resolution is wrong because the swapchain/Window size doesn't sync with Present. I think we have to record a size change and wait until next Present() to apply it. Probably also separate execution paths that end up to CreateSizeDependentResources based on whether the change comes from a rotation (external) or from the code (internal).
The following code swaps between 1 & 1/2 scale every frame.

> ```
>        f = (f+1)%2;
>        float scale = (f + 1f) / 1f; // swap between x1.0 or x0.5
>        graphics.PreferredBackBufferWidth = (int)(Window.ClientBounds.Width * scale);
>        graphics.PreferredBackBufferHeight = (int)(Window.ClientBounds.Height * scale);
>        graphics.ApplyChanges();
> ```

The first screenshot is half the size while it should scale up and fill the window and look  low res.
The opposite is also true, if you start with x0.5 scale and change to x1 the next frame is the upper-left 1/4 of your game zoomed in.
The second screenshot is Photoshoped to show how the game looks like, in realilty there is flicker.

![scaletohalf](https://cloud.githubusercontent.com/assets/3018589/7123997/ba86adf6-e22e-11e4-812d-0ff77ee67545.jpg)
![psed](https://cloud.githubusercontent.com/assets/3018589/7124221/0b56f5e6-e230-11e4-9365-bf7ac2f2d13e.jpg)

**Comment by tomspilman** [comment]
> The first frame after changing resolution is wrong because the swapchain/Window size doesn't sync with Present.

I suspect this has more to do with when in your own code you do the resolution change.  Are you calling `ApplyChanges()` during update or draw?

**Comment by nkast** [comment]
At the very beginning of `Draw()`. When I move it to the end, after `base.Draw()`, the result is a black screen. If I place it in `Updade()` the result is the same.

I made a mistake above. The code swap between x2 and x1. I changed the code to
> float scale = 1f/ (f + 1f);  // swap between x0.5 and x1.0
which produce the following 
![scaletohalf2](https://cloud.githubusercontent.com/assets/3018589/7213273/f91b02e2-e580-11e4-9b40-19f6ad11bfa6.jpg)

I am not quite sure what's going on here. Perhaps it's related to XAML. In the first frame we resize the backbuffer and the swapchain to half but for some reason it ends at twice as big in the screen until the next frame/present().

**Comment by nkast** [comment]
The platformer sample is not the best to demonstrate the problem because it doesn't adapt to resolution change, always render at 800x480. Here is photo of the MonoGame template. 
![wp_20150925_001 - copy](https://cloud.githubusercontent.com/assets/3018589/10100043/9f386da6-6397-11e5-9737-a03684562cdd.jpg)

I tried a couple of things, like moving swapchain or SwapChainBackgroundPanel res change before or after present(). I realized the problem is that SwapChainBackgroundPanel/swapchain update is depended on the XAML update cycle. I tried to force it by calling several xaml methods like .UpdateLayout() , .InvalidateMeasure(), nothing solved it.

 Then I found this (scroll down to "SwapChainPanel: Size Synchronization"):  https://blogs.windows.com/buildingapps/2013/11/19/whats-new-for-xaml-and-directx-interop-in-windows-8-1/ 
Maybe we can have better result with SwapChainPanel but right now we only use it on Win10.
